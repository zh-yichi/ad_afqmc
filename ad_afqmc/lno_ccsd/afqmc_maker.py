from jax import numpy as jnp
from jax import vmap

def walker_norhf_energy(walker,ham_data,trial_mo):
    '''rhf walker's energy with non-orthonal trial'''
    rot_h1,rot_chol = ham_data['rot_h1'],ham_data['rot_chol']
    nocc = rot_h1.shape[0]
    green_walker = (walker.dot(jnp.linalg.inv(trial_mo.T.conj() @ walker))).T
    f = jnp.einsum('gia,ak->gik',rot_chol[:,:nocc,nocc:],
                   green_walker.T[nocc:,:nocc],optimize='optimal')
    tf = vmap(jnp.trace)(f)
    tc = vmap(jnp.trace)(rot_chol[:,:nocc,:nocc])
    ene11 = 2*jnp.einsum('ia,ia->',green_walker[:nocc,nocc:],rot_h1[:nocc,nocc:])
    ene12 = 2*jnp.sum(tf*tc) \
            - jnp.einsum('gik,gki->',f,rot_chol[:,:nocc,:nocc])
    ene1 = ene11-2*ene12
    ene2 = 2*jnp.sum(tf*tf) \
           - jnp.einsum('gij,gji->',f,f)
    return ene1+ene2 #,ene1,ene2,

def walker_norhf_orb_en(walker,ham_data,wave_data):
    '''rhf walker's energy with non-orthonal trial'''
    rot_h1,rot_chol = ham_data['rot_h1'],ham_data['rot_chol']
    m = jnp.dot(wave_data["prjlo"].T,wave_data["prjlo"])
    nocc = rot_h1.shape[0]
    green_walker = (walker.dot(jnp.linalg.inv(
        wave_data['mo_coeff'].T.conj() @ walker))).T
    f = jnp.einsum('gia,ak->gik',rot_chol[:,:nocc,nocc:],
                   green_walker.T[nocc:,:nocc],optimize='optimal')
    tf = vmap(jnp.trace)(f)
    # tc = vmap(jnp.trace)(rot_chol[:,:nocc,:nocc])
    ene11 = 2*jnp.einsum('ia,ka,ik->',green_walker[:nocc,nocc:],rot_h1[:nocc,nocc:],m)
    ene12 = 2*jnp.einsum('gik,ik,gjj->',f,m,rot_chol[:,:nocc,:nocc]) \
            - jnp.einsum('gij,gjk,ik->',f,rot_chol[:,:nocc,:nocc],m)
    ene1 = ene11-2*ene12
    ene2 = 2*jnp.einsum('gik,ik,g->',f,m,tf) \
           - jnp.einsum('gij,gjk,ik->',f,f,m)
    return ene1+ene2