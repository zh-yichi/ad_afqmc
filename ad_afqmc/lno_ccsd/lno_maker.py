import numpy as np

def thouless_trans(t1):
    q, r = np.linalg.qr(t1)
    u_ai = r.T
    u_ji = q
    u_occ = np.vstack((u_ji,u_ai))
    u, _, _ = np.linalg.svd(u_occ)
    return u
