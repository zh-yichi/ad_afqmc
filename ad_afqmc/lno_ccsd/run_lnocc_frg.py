from functools import partial
from jax import random
#from mpi4py import MPI
import numpy as np
from jax import numpy as jnp
from ad_afqmc import config, wavefunctions, stat_utils
from ad_afqmc.lno_ccsd import lno_ccsd
import time

print = partial(print, flush=True)

ham_data, ham, prop, trial, wave_data, sampler, observable, options, _ = (
    lno_ccsd.prep_lnoccsd_afqmc())

if options["use_gpu"]:
    config.afqmc_config["use_gpu"] = True

config.setup_jax()
MPI = config.setup_comm()

# init_time = time.time()
comm = MPI.COMM_WORLD
rank = comm.Get_rank()  # Process rank
size = comm.Get_size()  # Total number of processes


### initialize propagation
seed = options["seed"]
propagator = prop
init_walkers = None
ham_data = wavefunctions.rhf(trial.norb, trial.nelec,n_batch=trial.n_batch
                                )._build_measurement_intermediates(ham_data, wave_data)
ham_data = ham.build_measurement_intermediates(ham_data, trial, wave_data)
ham_data = ham.build_propagation_intermediates(ham_data, propagator, trial, wave_data)

prop_data = propagator.init_prop_data(trial, wave_data, ham_data, init_walkers)
if jnp.abs(jnp.sum(prop_data["overlaps"])) < 1.0e-6:
    raise ValueError(
        "Initial overlaps are zero. Pass walkers with non-zero overlap."
    )
prop_data["key"] = random.PRNGKey(seed + rank)

prop_data["overlaps"] = trial.calc_overlap(prop_data["walkers"], wave_data)
prop_data["n_killed_walkers"] = 0
prop_data["pop_control_ene_shift"] = prop_data["e_estimate"]

### relaxation ###
init_time = time.time()
comm.Barrier()
if rank == 0:
    hf_orb_cr,hf_orb_en = lno_ccsd._frg_hf_cr(
        ham_data['rot_h1'], ham_data['rot_chol'],prop_data["walkers"][0],trial,wave_data)
    cc_orb_cr = lno_ccsd._frg_ccsd_cr(
        prop_data["walkers"][0],ham_data,wave_data,trial,1e-6)
    
    cc_orb_en = hf_orb_cr+cc_orb_cr
    e_init = prop_data["e_estimate"]

    print(f'# afqmc propagation with {options["n_walkers"]*size} walkers')
    print(f'# appraching equilibrium')
    print('# step  energy  hf_orb_en  cc_orb_en  time')
    print(f"  {0:3d}"
          f"  {e_init:.6f}"
          f"  {hf_orb_en:.6f}"
          f"  {cc_orb_en:.6f}"
          f"  {time.time() - init_time:.2f} "
        )
comm.Barrier()

for n in range(options["n_eql"]):
    prop_data, (blk_en,blk_wt,blk_hf_orb_en,blk_cc_orb_en) \
                = lno_ccsd.propagate_phaseless_orb(
                    ham_data,prop,prop_data,trial,wave_data,sampler)
    
    blk_en = np.array([blk_en], dtype="float32")
    blk_wt = np.array([blk_wt], dtype="float32")
    blk_hf_orb_en = np.array([blk_hf_orb_en], dtype="float32")
    blk_cc_orb_en = np.array([blk_cc_orb_en], dtype="float32")    

    blk_wt_en = np.array(
        [blk_en * blk_wt], dtype="float32"
    )
    blk_wt_hf_orb_en = np.array(
        [blk_hf_orb_en * blk_wt], dtype="float32"
    )
    blk_wt_cc_orb_en = np.array(
        [blk_cc_orb_en * blk_wt], dtype="float32"
    )

    tot_wt_en = np.zeros(1, dtype="float32")
    tot_wt = np.zeros(1, dtype="float32")
    tot_wt_hf_orb_en = np.zeros(1, dtype="float32")
    tot_wt_cc_orb_en = np.zeros(1, dtype="float32")

    comm.Reduce(
            [blk_wt_en, MPI.FLOAT],
            [tot_wt_en, MPI.FLOAT],
            op=MPI.SUM,
            root=0,
        )
    comm.Reduce(
        [blk_wt, MPI.FLOAT],
        [tot_wt, MPI.FLOAT],
        op=MPI.SUM,
        root=0,
    )
    comm.Reduce(
            [blk_wt_hf_orb_en, MPI.FLOAT],
            [tot_wt_hf_orb_en, MPI.FLOAT],
            op=MPI.SUM,
            root=0,
        )
    comm.Reduce(
            [blk_wt_cc_orb_en, MPI.FLOAT],
            [tot_wt_cc_orb_en, MPI.FLOAT],
            op=MPI.SUM,
            root=0,
        )

    comm.Barrier()
    if rank == 0:
        blk_en = tot_wt_en / tot_wt
        blk_wt = tot_wt
        blk_hf_orb_en = tot_wt_hf_orb_en / tot_wt
        blk_cc_orb_en = tot_wt_cc_orb_en / tot_wt

    comm.Bcast(blk_en, root=0)
    comm.Bcast(blk_wt, root=0)
    comm.Bcast(blk_hf_orb_en, root=0)
    comm.Bcast(blk_cc_orb_en, root=0)
    
    prop_data = propagator.orthonormalize_walkers(prop_data)
    prop_data = propagator.stochastic_reconfiguration_global(prop_data, comm)
    prop_data["e_estimate"] = (
         0.9 * prop_data["e_estimate"] + 0.1 * blk_en[0]
         )
    comm.Barrier()

    comm.Barrier()
    if rank == 0:
        print(
            f"  {n+1:3d}"
            f"  {blk_en[0]:.6f}"
            f"  {blk_hf_orb_en[0]:.6f}"
            f"  {blk_cc_orb_en[0]:.6f}"
            f"  {time.time() - init_time:.2f} "
        )
    comm.Barrier()

glb_blk_en = None
glb_blk_wt = None
glb_blk_hf_orb_en = None
glb_blk_cc_orb_en = None

if rank == 0:
    glb_blk_en = np.zeros(size * sampler.n_blocks)
    glb_blk_wt = np.zeros(size * sampler.n_blocks)
    glb_blk_hf_orb_en = np.zeros(size * sampler.n_blocks)
    glb_blk_cc_orb_en = np.zeros(size * sampler.n_blocks)
    
comm.Barrier()
if rank == 0:
    print("# Sampling sweeps:")
    print("# iter  energy  err  hf_orb_en  err  cc_orb_cr  err  time")
comm.Barrier()

for n in range(sampler.n_blocks):
    prop_data,(blk_en,blk_wt,blk_hf_orb_en,blk_cc_orb_en) \
        = lno_ccsd.propagate_phaseless_orb(
            ham_data,prop,prop_data,trial,wave_data,sampler)

    blk_en = np.array([blk_en], dtype="float32")
    blk_wt = np.array([blk_wt], dtype="float32")
    blk_hf_orb_en = np.array([blk_hf_orb_en], dtype="float32")
    blk_cc_orb_en = np.array([blk_cc_orb_en], dtype="float32")

    gather_en = None
    gather_wt = None
    gather_hf_orb_en = None
    gather_cc_orb_en = None

    comm.Barrier()
    if rank == 0:
        gather_en = np.zeros(size, dtype="float32")
        gather_wt = np.zeros(size, dtype="float32")
        gather_hf_orb_en = np.zeros(size, dtype="float32")
        gather_cc_orb_en = np.zeros(size, dtype="float32")
    comm.Barrier()

    comm.Gather(blk_en, gather_en, root=0)
    comm.Gather(blk_wt, gather_wt, root=0)
    comm.Gather(blk_hf_orb_en, gather_hf_orb_en, root=0)
    comm.Gather(blk_cc_orb_en, gather_cc_orb_en, root=0)

    comm.Barrier()
    if rank == 0:
        glb_blk_en[n * size : (n + 1) * size] = gather_en
        glb_blk_wt[n * size : (n + 1) * size] = gather_wt
        glb_blk_hf_orb_en[n * size : (n + 1) * size] = gather_hf_orb_en
        glb_blk_cc_orb_en[n * size : (n + 1) * size] = gather_cc_orb_en

        assert gather_wt is not None

        blk_wt= np.sum(gather_wt)
        blk_en = np.sum(gather_wt * gather_en) / blk_wt
    comm.Barrier()
    
    prop_data = propagator.orthonormalize_walkers(prop_data)
    prop_data = propagator.stochastic_reconfiguration_global(prop_data, comm)
    prop_data["e_estimate"] = 0.9 * prop_data["e_estimate"] + 0.1 * blk_en

    if n % (max(sampler.n_blocks // 10, 1)) == 0:
        comm.Barrier()
        if rank == 0:
                en, en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_en[: (n + 1) * size],
                    neql=0,
                )
                hf_orb_en, hf_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_hf_orb_en[: (n + 1) * size],
                    neql=0,
                )
                cc_orb_en, cc_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_cc_orb_en[: (n + 1) * size],
                    neql=0,
                )

                if en_err is not None:
                    en_err = f"{en_err:.6f}"
                else:
                    en_err = f"  {en_err}  "
                if hf_orb_en_err is not None:
                    hf_orb_en_err = f"{hf_orb_en_err:.6f}"
                else:
                    hf_orb_en_err = f"  {hf_orb_en_err}  "
                if cc_orb_en_err is not None:
                    cc_orb_en_err = f"{cc_orb_en_err:.6f}"
                else:
                    cc_orb_en_err = f"  {cc_orb_en_err}  "

                en = f"{en:.6f}"
                hf_orb_en = f"{hf_orb_en:.6f}"
                cc_orb_en = f"{cc_orb_en:.6f}"

                print(f"  {n:4d}  {en}  {en_err}  {hf_orb_en}  {hf_orb_en_err}"
                      f"  {cc_orb_en}  {cc_orb_en_err}  {time.time() - init_time:.2f}")
        comm.Barrier()


comm.Barrier()
if rank == 0:
    assert glb_blk_en is not None
    assert glb_blk_wt is not None
    assert glb_blk_hf_orb_en is not None
    assert glb_blk_cc_orb_en is not None


    samples_clean, idx = stat_utils.reject_outliers(
    np.stack(
                (
                    glb_blk_wt,
                    glb_blk_en,
                    glb_blk_hf_orb_en,
                    glb_blk_cc_orb_en,
                )
            ).T,
            1,
        )

    print(
        f"# Number of outliers in post: {glb_blk_wt.size - samples_clean.shape[0]} "
        )
    
    glb_blk_wt = samples_clean[:, 0]
    glb_blk_en = samples_clean[:, 1]
    glb_blk_hf_orb_en = samples_clean[:, 2]
    glb_blk_cc_orb_en = samples_clean[:, 3]

    en, en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_en[: (n + 1) * size],
                    neql=0,printQ=True
                )
    hf_orb_en, hf_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_hf_orb_en[: (n + 1) * size],
                    neql=0,printQ=True
                )
    cc_orb_en, cc_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_cc_orb_en[: (n + 1) * size],
                    neql=0,printQ=True
                )
    
    if en_err is not None:
        en_err = f"{en_err:.6f}"
    else:
        en_err = f"  {en_err}  "
    if hf_orb_en_err is not None:
        hf_orb_en_err = f"{hf_orb_en_err:.6f}"
    else:
        hf_orb_en_err = f"  {hf_orb_en_err}  "
    if cc_orb_en_err is not None:
        cc_orb_en_err = f"{cc_orb_en_err:.6f}"
    else:
        cc_orb_en_err = f"  {cc_orb_en_err}  "


    en = f"{en:.6f}"
    hf_orb_en = f"{hf_orb_en:.6f}"
    cc_orb_en = f"{cc_orb_en:.6f}"

    print(f"Final Results:")
    print(f"lno-ccsd-afqmc energy: {en} +/- {en_err}")
    print(f"lno-ccsd-afqmc hf_orb_en: {hf_orb_en} +/- {hf_orb_en_err}")
    print(f"lno-ccsd-afqmc cc_orb_en: {cc_orb_en} +/- {cc_orb_en_err}")
    print(f"total run time: {time.time() - init_time:.2f}")

comm.Barrier()
