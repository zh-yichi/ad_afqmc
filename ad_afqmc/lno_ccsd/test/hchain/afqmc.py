from functools import partial
from pyscf import gto, scf, cc
from ad_afqmc import pyscf_interface, run_afqmc

from jax import config
config.update("jax_enable_x64", True)

print = partial(print, flush=True)

a = 1.
nH = 4
atoms = ""
for i in range(nH):
    atoms += f"H {i*a} 0 0 \n"

mol = gto.M(atom=atoms, basis="sto6g", verbose=4)
mf = scf.RHF(mol).density_fit()
mf.kernel()

# cc
mycc = cc.CCSD(mf)
mycc.kernel()
et = mycc.ccsd_t()
print(f"ccsd energy is {mycc.e_tot}")
print(f"ccsd_t energy is {mycc.e_tot+et}")


options = {
    "n_eql": 4,
    "n_ene_blocks": 1,
    "n_sr_blocks": 20,
    "n_blocks": 40,
    "n_walkers": 50,
    "seed": 98,
    "trial": "cisd",
    "walker_type": "rhf",
    "dt":0.005,
}


mo_coeff = mf.mo_coeff
nfrozen = 0
norb_act = mol.nao - nfrozen
nelec_act = mol.nelectron - 2 * nfrozen
t1 = mycc.t1
t2 = mycc.t2


from ad_afqmc.lno_ccsd import lno_ccsd
lno_ccsd.prep_lno_amp_chol_file(mf,mo_coeff,options,norb_act=norb_act,nelec_act=nelec_act,
                                prjlo=[],norb_frozen=nfrozen,t1=t1,t2=t2,
                                df=True,chol_cut=1e-6,
                                option_file='options.bin',
                                mo_file="mo_coeff.npz",
                                amp_file="amplitudes.npz",
                                chol_file="FCIDUMP_chol"
                                )

from mpi4py import MPI

MPI.Finalize()
run_afqmc.run_afqmc(options=options, nproc=5)
