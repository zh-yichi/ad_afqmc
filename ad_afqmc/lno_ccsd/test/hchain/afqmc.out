#INFO: **** input file is /home/yichi/research/software/cs_afqmc/ad_afqmc/lno_ccsd/test/hchain/afqmc.py ****
from functools import partial
from pyscf import gto, scf, cc
from ad_afqmc import pyscf_interface, run_afqmc

from jax import config
config.update("jax_enable_x64", True)

print = partial(print, flush=True)

a = 1.
nH = 6
atoms = ""
for i in range(nH):
    atoms += f"H {i*a} 0 0 \n"

mol = gto.M(atom=atoms, basis="sto6g", verbose=4)
mf = scf.RHF(mol).density_fit()
mf.kernel()

# cc
mycc = cc.CCSD(mf)
mycc.kernel()
et = mycc.ccsd_t()
print(f"ccsd energy is {mycc.e_tot}")
print(f"ccsd_t energy is {mycc.e_tot+et}")

options = {
    "n_eql": 4,
    "n_ene_blocks": 1,
    "n_sr_blocks": 20,
    "n_blocks": 40,
    "n_walkers": 50,
    "seed": 98,
    "trial": "cisd",
    "walker_type": "rhf",
    "dt":0.005,
}

# mo_coeff = mf.mo_coeff
# nfrozen = 0
# norb_act = mol.nao - nfrozen
# nelec_act = mol.nelectron - 2 * nfrozen
# t1 = mycc.t1
# t2 = mycc.t2

# from ad_afqmc.lno_ccsd import lno_ccsd
# lno_ccsd.prep_lno_amp_chol_file(mf,mo_coeff,options,norb_act=norb_act,nelec_act=nelec_act,
#                                 prjlo=[],norb_frozen=nfrozen,t1=t1,t2=t2,
#                                 df=True,chol_cut=1e-6,
#                                 option_file='options.bin',
#                                 mo_file="mo_coeff.npz",
#                                 amp_file="amplitudes.npz",
#                                 chol_file="FCIDUMP_chol"
#                                 )

pyscf_interface.prep_afqmc(mycc,use_df=True)

from mpi4py import MPI

MPI.Finalize()
run_afqmc.run_afqmc(options=options, nproc=5)
#INFO: ******************** input file end ********************


System: uname_result(system='Linux', node='yichi-thinkpad', release='4.4.0-26100-Microsoft', version='#1882-Microsoft Fri Jan 01 08:00:00 PST 2016', machine='x86_64')  Threads 12
Python 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
numpy 1.24.3  scipy 1.14.1  h5py 3.12.1
Date: Fri Jul 18 21:57:25 2025
PySCF version 2.8.0
PySCF path  /home/yichi/research/software/lno_pyscf
GIT HEAD (branch master) ef75f4190e4de208685670651dc6c467f72b6794

[ENV] PYSCF_EXT_PATH /home/yichi/research/software/pyscf
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 6
[INPUT] num. electrons = 6
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry False subgroup None
[INPUT] Mole.unit = angstrom
[INPUT] Symbol           X                Y                Z      unit          X                Y                Z       unit  Magmom
[INPUT]  1 H      0.000000000000   0.000000000000   0.000000000000 AA    0.000000000000   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  2 H      1.000000000000   0.000000000000   0.000000000000 AA    1.889726124565   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  3 H      2.000000000000   0.000000000000   0.000000000000 AA    3.779452249130   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  4 H      3.000000000000   0.000000000000   0.000000000000 AA    5.669178373695   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  5 H      4.000000000000   0.000000000000   0.000000000000 AA    7.558904498260   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  6 H      5.000000000000   0.000000000000   0.000000000000 AA    9.448630622825   0.000000000000   0.000000000000 Bohr   0.0

nuclear repulsion = 4.603841735004
number of shells = 6
number of NR pGTOs = 36
number of NR cGTOs = 6
basis = sto6g
ecp = {}
CPU time:         0.58


******** <class 'pyscf.df.df_jk.DFRHF'> ********
method = DFRHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
diis_damp = 0
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 50
direct_scf = False
chkfile to save SCF result = /tmp/tmpzgududdq
max_memory 4000 MB (current use 133 MB)
Set gradient conv threshold to 3.16228e-05
Initial guess from minao.
******** <class 'pyscf.df.df.DF'> ********
auxbasis = None
max_memory = 4000
ETB for H: l = 0, exps = 0.200224856 * 2^n , n = 0..8

WARN: Even tempered Gaussians are generated as DF auxbasis for  H

init E= -2.42992244139156
  HOMO = -0.204045735961534  LUMO = 0.13840111927303
cycle= 1 E= -3.14955212808499  delta_E= -0.72  |g|= 0.111  |ddm|= 1.71
  HOMO = -0.324221355929662  LUMO = 0.214034118577356
cycle= 2 E= -3.15774245178069  delta_E= -0.00819  |g|= 0.0385  |ddm|= 0.246
  HOMO = -0.330464366202016  LUMO = 0.232009176330945
cycle= 3 E= -3.1589679408949  delta_E= -0.00123  |g|= 0.00665  |ddm|= 0.101
  HOMO = -0.333957770303782  LUMO = 0.232352367960062
cycle= 4 E= -3.15900173729128  delta_E= -3.38e-05  |g|= 0.000168  |ddm|= 0.0212
  HOMO = -0.333844822767501  LUMO = 0.232284255609339
cycle= 5 E= -3.15900175306578  delta_E= -1.58e-08  |g|= 1.37e-05  |ddm|= 0.000402
  HOMO = -0.333861918151937  LUMO = 0.232293175564539
cycle= 6 E= -3.15900175322303  delta_E= -1.57e-10  |g|= 1.22e-07  |ddm|= 4.33e-05
  HOMO = -0.333862067839926  LUMO = 0.232293225293638
Extra cycle  E= -3.15900175322304  delta_E= -7.99e-15  |g|= 4.17e-08  |ddm|= 2.19e-07
converged SCF energy = -3.15900175322304

******** <class 'pyscf.cc.dfccsd.RCCSD'> ********
CC2 = 0
CCSD nocc = 3, nmo = 6
max_cycle = 50
direct = 0
conv_tol = 1e-07
conv_tol_normt = 1e-05
diis_space = 6
diis_start_cycle = 0
diis_start_energy_diff = 1e+09
max_memory 4000 MB (current use 139 MB)
Init t2, MP2 energy = -3.22148731886451  E_corr(MP2) -0.0624855656414709
Init E_corr(RCCSD) = -0.0624855656414718
cycle = 1  E_corr(RCCSD) = -0.0844268325128873  dE = -0.0219412669  norm(t1,t2) = 0.0641505
cycle = 2  E_corr(RCCSD) = -0.0936927843715444  dE = -0.00926595186  norm(t1,t2) = 0.0336578
cycle = 3  E_corr(RCCSD) = -0.101894146972772  dE = -0.0082013626  norm(t1,t2) = 0.0180069
cycle = 4  E_corr(RCCSD) = -0.102078986082588  dE = -0.00018483911  norm(t1,t2) = 0.00416526
cycle = 5  E_corr(RCCSD) = -0.101955976758356  dE = 0.000123009324  norm(t1,t2) = 0.00148289
cycle = 6  E_corr(RCCSD) = -0.102028959806256  dE = -7.29830479e-05  norm(t1,t2) = 0.000320913
cycle = 7  E_corr(RCCSD) = -0.102008962955064  dE = 1.99968512e-05  norm(t1,t2) = 0.000100223
cycle = 8  E_corr(RCCSD) = -0.102011903771294  dE = -2.94081623e-06  norm(t1,t2) = 2.40446e-05
cycle = 9  E_corr(RCCSD) = -0.102012687116702  dE = -7.83345408e-07  norm(t1,t2) = 3.60363e-06
cycle = 10  E_corr(RCCSD) = -0.102012762597423  dE = -7.5480721e-08  norm(t1,t2) = 9.06956e-07
RCCSD converged
E(RCCSD) = -3.26101451582046  E_corr = -0.1020127625974229
RCCSD(T) correction = -0.000479676798005027
ccsd energy is -3.2610145158204595
ccsd_t energy is -3.2614941926184646
#
# Preparing AFQMC calculation
# If you import pyscf cc modules and use MPI for AFQMC in the same script, finalize MPI before calling the AFQMC driver.
# Calculating Cholesky integrals
# Decomposing ERI with DF
# Decomposing ERI with DF
# Finished calculating Cholesky integrals
#
# Size of the correlation space:
# Number of electrons: (3, 3)
# Number of basis functions: 6
# Number of Cholesky vectors: 54
#
# Hostname: yichi-thinkpad
# System Type: Linux
# Machine Type: x86_64
# Processor: x86_64
# Number of MPI ranks: 5
#
# norb: 6
# nelec: (3, 3)
#
# n_eql: 4
# n_ene_blocks: 1
# n_sr_blocks: 20
# n_blocks: 40
# n_walkers: 50
# seed: 98
# trial: cisd
# walker_type: rhf
# dt: 0.005
# n_exp_terms: 6
# n_prop_steps: 50
# orbital_rotation: True
# do_sr: True
# symmetry: False
# save_walkers: False
# ene0: 0.0
# free_projection: False
# n_batch: 1
# LNO: False
# orbE: 0
# maxError: 0.001
#
# Equilibration sweeps:
#   Iter        Block energy      Walltime
#     0      -3.261014506e+00     7.77e+00 
#     1      -3.261695147e+00     2.86e+01 
#     2      -3.261366367e+00     4.73e+01 
#     3      -3.261397839e+00     5.56e+01 
#     4      -3.261339188e+00     6.38e+01 
#
# Sampling sweeps:
#  Iter        Mean energy          Stochastic error       Walltime
     0      -3.261851904e+00        4.188184307e-04        7.68e+01 
     4      -3.261601740e+00        1.598016681e-04        9.06e+01 
     8      -3.261572659e+00        1.090433914e-04        1.04e+02 
    12      -3.261444850e+00        8.514113664e-05        1.19e+02 
    16      -3.261427748e+00        7.087461152e-05        1.33e+02 
    20      -3.261388785e+00        6.061894458e-05        1.47e+02 
    24      -3.261353879e+00        5.848419534e-05        1.60e+02 
    28      -3.261367836e+00        5.223501647e-05        1.74e+02 
    32      -3.261357835e+00        4.908570752e-05        1.88e+02 
    36      -3.261335225e+00        4.968535803e-05        2.01e+02 
#
# Number of large deviations: 0
# Number of outliers in post: 0 
#
# Mean: -3.26137910e+00
# Block size    # of blocks         Mean                Error
      1              200       -3.26137910e+00       5.012106e-05
      2              100       -3.26137910e+00       5.286687e-05
      5               40       -3.26137910e+00       5.135147e-05
     10               20       -3.26137910e+00       6.118593e-05
     20               10       -3.26137910e+00       6.914484e-05
     50                4       -3.26137910e+00       6.477584e-05
# Stocahstic error estimate: 5.286687e-05
#
AFQMC energy: -3.26138 +/- 0.00005

ph_afqmc walltime: 211.6885209083557
